<a href="https://colab.research.google.com/github/alexDx12/gb_statistics/blob/lesson_1/lesson_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Основные задачи

### Задача 1

#### Постановка задачи

Имеется колода в 52 карты. Найти число возможностей вытянуть из неё 4 карты так, чтобы среди них был хотя бы один туз.

#### Решение

Импорт библиотек и скриптов:

In [ ]:
import numpy as np

Разделим колоду на 2 части - в первую часть отложим четыре туза, а во вторую - оставшиеся карты. После этого определим общее количество сочетаний для каждого случая количества выбранных тузов.

Массив сочетаний карт из первой части колоды (4 карты):

In [ ]:
C1 = np.array([np.math.factorial(4) // (np.math.factorial(k1) * np.math.factorial(4 - k1)) for k1 in range(1, 5)])
C1

array([4, 6, 4, 1])

Массив сочетаний карт из второй части колоды (48 карт):

In [ ]:
C2 = np.array([np.math.factorial(48) // (np.math.factorial(4 - k2) * np.math.factorial(48 - (4 - k2))) for k2 in range(1, 5)])
C2

array([17296,  1128,    48,     1])

Массив сочетаний карт из обоих частей колоды:

In [ ]:
C = C1 * C2
C

array([69184,  6768,   192,     1])

Общее количество всех сочетаний карт:

In [ ]:
sum(C)

76145

Ответ: существует 76145 возможностей вытянуть из колоды 4 карты так, чтобы среди них был хотя бы один туз.

### Задачи 2 и 3

#### Постановка задач

Из 60 вопросов, входящих в экзаменационные билеты, студент знает 50. Случайным образом студент вытягивает 3 вопроса. Какова вероятность, что все выбранные вопросы знакомы студенту? Какова вероятность что два из трёх вопросов знакомы студенту?

#### Решение

Импорт библиотек и скриптов:

In [ ]:
import numpy as np

Общее количество исходов при вытягивании трёх вопросов из шестидесяти:

In [ ]:
n1 = np.math.factorial(60) // (np.math.factorial(3) * np.math.factorial(60 - 3))
n1

34220

Количество исходов при вытягивании трёх знакомых студенту вопросов из пятидесяти: 

In [ ]:
n2 = np.math.factorial(50) // (np.math.factorial(3) * np.math.factorial(50 - 3))
n2

19600

Вероятность вытягивания трёх знакомых студенту вопросов из шестидесяти:

In [ ]:
p1 = n2 / n1
p1

0.5727644652250146

In [ ]:
round(p1, 2)

0.57

Количество исходов при вытягивании двух знакомых студенту вопросов:

In [ ]:
n3 = np.math.factorial(50) // (np.math.factorial(2) * np.math.factorial(50 - 2))
n3

1225

Количество исходов при вытягивании одного не знакомого студенту вопроса:

In [ ]:
n4 = np.math.factorial(10) // (np.math.factorial(1) * np.math.factorial(10 - 1))
n4

10

Количество исходов при одновременном вытягивании двух знакомых и одного не знакомого студенту вопроса:

In [ ]:
n5 = n3 * n4
n5

12250

Вероятность вытягивания двух знакомых и одного не знакомого студенту вопроса:

In [ ]:
p2 = n5 / n1
p2

0.35797779076563413

In [ ]:
round(p2, 2)

0.36

Ответ: вероятность того, что все выбранные вопросы знакомы студенту равяется 0.57, вероятность того, что два из трёх вопросов знакомы студенту равняется 0.36.

### Задача 4

#### Постановка задачи

Допустим, имеется некоторая очень редкая болезнь (поражает 0.1% населения). Вы приходите к врачу, вам делают тест на эту болезнь, и тест оказывается положительным. Врач говорит вам, что этот тест верно выявляет 99% больных этой болезнью и всего лишь в 1% случаев даёт ложный положительный ответ.

Вопрос: какова вероятность, что вы действительно больны ей?

Подсказка: используйте формулу Байеса с раскрытием знаменателя с помощью формулы полной вероятности.

#### Решение

Определим следующие события. Событие $A$ — вы больны этой болезнью, событие $B$ — для вас тест на эту болезнь оказался положительным.
В этом случае нам нужно найти вероятность $P(A | B)$ (поскольку вы уже знаете, что тест дал положительный результат). По формуле Байеса:
$$P(A|B) = \frac{P(B \cap A)}{P(B)} = \frac{P(B | A) \cdot P(A)}{P(B)}$$

Также раскроем знаменатель по формуле полной вероятности следующим образом:
$$P(B) = P(B | A) \cdot P(A) + P \left( B | \overline{A} \right) \cdot P \left( \overline{A} \right)$$

In [ ]:
P_b = 0.001 * 0.99 + 0.999 * 0.01
P_b

0.01098

In [ ]:
P_a_b = (0.001 * 0.99) / (0.99 * 0.001 + 0.01 * 0.999)
round(P_a_b, 2)

0.09

Ответ: вероятность болезни составляет 0.09.

## Дополнительные задачи

### Задача 1

#### Постановка задачи

Подбрасываются три игральных кубика. Оцените вероятность того, что сумма значений этих кубиков не будет превышать 6.

_Подсказка_: оценкой вероятности может служить, например, относительная частота при большом числе опытов. Попробуйте реализовать наибольшее число опытов, которое позволяют ваш компьютер и/или ваше терпение.

_Замечание_: вообще в этом курсе мы не используем такие численные методы оценки вероятностей, и в будущем мы не будем их применять. Но тут сделаем исключение один раз.

#### Решение

In [ ]:
import random

Найдём количесто реализаций, при которых сумма значений кубиков не превысит 6 (m) при подбрасывании их определённое количество раз (n). Ограничение количества подбрасываний задаим с помощью переменной limit.

In [ ]:
limit = 10000000
m = 0
n = 0
while n < limit:
    dice_1 = random.randint(1, 6)
    dice_2 = random.randint(1, 6)
    dice_3 = random.randint(1, 6)
    dice_sum = dice_1 + dice_2 + dice_3
    if dice_sum <= 6:
        m += 1
    n += 1

Количесто реализаций, при которых сумма значений кубиков не превысит 6:

In [ ]:
m

925238

Количество подбрасываний кубиков:

In [ ]:
n

10000000

Искомая веротность:

In [ ]:
P = m/n
round(P, 4)

0.0925

Ответ: вероятность того, что сумма значений этих кубиков не будет превышать 6 составляет: 0,0927.

### Задача 2

#### Постановка задачи

Бросается игральная кость. Пусть событие `A` - появление чётного числа, событие `B` - появление числа больше трёх. Являются ли эти события независимыми? Почему?

#### Решение

Вероятность события A при условии наступления события B (выпадение чётных чисел 4 и 6 срдеи чисел больше трёх - 4, 5 и 6):

In [ ]:
2 / 3

0.6666666666666666

Вероятность события A (выпадение чётных чисел 2, 4 и 6 из остальных - 1, 2, 3, 4, 5 и 6):

In [ ]:
3 / 6

0.5

По определению события $A$ и $B$ являются независимыми, если $P(A|B) = P(A)$. Поскольку в рассмотренном случае $P(A|B) \neq P(A)$, то события A и B являются зависимыми.

Ответ: события A и B являются зависимыми, т.к. $P(A|B) = P(A)$.